test/playwright/test_react_router.py

```diff
@@ -0,0 +1,88 @@
+"""Test app with different types of React routers
+
+- browser router (uses the history API)
+- memory router (uses in-app memory)
+
+Ref: https://reactrouter.com/en/main/routers/create-memory-router
+"""
+
+import pytest
+import re
+from playwright.sync_api import Page, expect
+
+
+@pytest.fixture
+def test_config():
+    return {"base_url": "http://localhost:8000"}
+
+
+def test_browser_router_200_ok(page: Page, test_config):
+    """With browser router, a known route should show the corresponding view
+    """
+    # Check that when going to a known route (in this case, the route to create
+    # a new environment), the app loads the view for that route.
+    page.goto(test_config["base_url"] + "/default/new-environment")
+
+    # We know we are at the correct view (i.e., new environment form) if there
+    # is a textbox to enter the name of the new environment.
+    expect(page.get_by_role("textbox", name="environment name")).to_be_visible()
+
+
+def test_memory_router_200_ok():
+    """With memory router, all routes are 200 (OK) so there's nothing to test there
+    """
+    pass
+
+
+def test_browser_router_404_not_found(page: Page, test_config):
+    """With browser router, an unknown route should result in a 404 not found error
+    """
+    page.goto(test_config["base_url"] + "/this-is-not-an-app-route")
```

Comment:
Oh, hm, yeah, that is confusing. 

I have actually forgotten best practices for this. 

I recently created a related conda-store [server issue about creating a catch-all route that returns the UI app instead of a 404](https://github.com/conda-incubator/conda-store/issues/899), if no other route is matched.

I will have to do some research, but I think that ultimately if we want an actual HTTP 404 status code, we will have to define the UI app routes in both the client-side and server-side code.

The only reason these nonsense routes, such as /foo-bar, return the UI app with a 200, by the way, is because of the [`--history-api-fallback` Webpack dev server flag](https://webpack.js.org/configuration/dev-server/#devserverhistoryapifallback) passed to the package.json `start` script:

```
"start:ui": "REACT_APP_VERSION=$npm_package_version webpack server --history-api-fallback",
```

Reply:
So the error code itself is intended behavior of `--history-api-fallback`. Is that good/what we want? Seems like invalid routes should return error codes.

Reply:
Hmm? I'm not sure I follow. In the example you pasted above, the `--history-api-fallback` flag is precisely why `/foo/bar` returns 200 (with the app bundle) instead of a 404. 

I did some web searching and I discovered that one solution (that doesn't require us to find a way to share routes between the front end and the back end) is to do a JavaScript redirect to a 404/not-found page when the user goes to an unknown route. Here's how that would work, step by step:

1. User visits /foo/bar
2. Server returns React app with 200 status code
3. React app boots up in the browser, doesn't recognize the /foo/bar route, executes client-side redirect JavaScript redirect like so: `window.location = "/not-found"`
4. Server returns a 404 page for the /not-found route. (Actually, it could maybe should return the React app at /not-found, but with a 404 status code instead of 200.)

Does that make sense? It's a little less than ideal. But the ideal solution requires us to either duplicate or share the React app routes between both the front-end and back-end codebases.